### PR TITLE
Reset MemoryCache during rebalancings

### PR DESCRIPTION
**Pull request by borzunov**
Before this PR, if there were open inference sessions right when rebalancing is triggered, their cache was never properly destroyed.

See the logs:

```
Dec 14 17:10:01.824 [INFO] rpc_inference.alloc(total_size=4.34 GiB), used 4.34/13.50 GiB (32.1%)      
Dec 14 17:11:16.547 [INFO] Swarm balance quality: 45.6%                                                                                                                                                     
Dec 14 17:11:16.547 [INFO] Swarm is imbalanced, server will load other blocks                      
Dec 14 17:11:16.863 [INFO] Announced that blocks [...] are offline                                                         
Dec 14 17:11:18.894 [INFO] Shutting down                                                              
Dec 14 17:11:18.898 [INFO] Module container shut down successfully                                                                                                                                          
Dec 14 17:11:18.987 [INFO] Cleaning up, left 110 open file descriptors                                                                                                                                      
Dec 14 17:11:19.052 [INFO] Cleaning up, left 4.3 GiB allocated memory, 11.2 GiB reserved memory
Dec 14 17:11:19.535 [INFO] Announced that blocks [43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69] are joining
...
Dec 14 18:07:19.325 [INFO] Started                                                                    
Dec 14 18:08:16.135 [INFO] Swarm balance quality: 78.5%          
Dec 14 18:09:07.562 [INFO] Swarm balance quality: 78.5%                                            
Dec 14 18:10:42.685 [INFO] Swarm balance quality: 78.5%          
Dec 14 18:11:57.418 [INFO] rpc_inference.open(blocks=43:70, remote_peer=...Q3aiDh)                 
Dec 14 18:11:57.427 [INFO] rpc_inference.alloc(total_size=3.78 GiB), used 8.12/13.50 GiB (60.2%)
```

Here, the first `alloc` in the ModuleContainer reports that 8.12 GiB = 4.34 GiB + 3.78 GiB are already allocated. The 4.34 GiB are actually left from the inference session opened before the rebalancing, and would be never freed without this fix.

### CODE REVIEW

File: src/petals/server/memory_cache.py [hunk 1]
```diff
@@ -33,12 +33,10 @@ def __init__(self, device: Union[str, torch.device], max_size_bytes: Optional[in
         self._lock_metadata, self.size_decreased_event = mp.Lock(), mp.Event()
         self._current_size = mp.Value(ctypes.c_int64, 0, lock=False)
         self._handle_counter = mp.Value(ctypes.c_int64, 0, lock=False)
-        self._active_handles: Optional[Dict[Handle, TensorDescriptor]] = None
-        self._allocated_tensors: Optional[Dict[Handle, torch.Tensor]] = None
+        self._allocated_tensors: Dict[Handle, torch.Tensor] = {}
```
Comment (borzunov) [inline src/petals/server/memory_cache.py:36]: Changes in this file just simplify code a bit without changing its functionality.

File: src/petals/server/server.py [hunk 2]
```diff
@@ -157,8 +157,8 @@ def __init__(
         if attn_cache_size is None:
             # Hidden size is 14336 for the bigscience/bloom-petals model. For other models, scale accordingly
             attn_cache_size = 0.5 * gib * num_blocks * self.block_config.hidden_size / 14336
+        self.attn_cache_size, self.alloc_timeout = attn_cache_size, alloc_timeout
```
Comment (borzunov) [inline src/petals/server/server.py:160]: Now, we will create `MemoryCache` only in `ModuleContainer`, so it's garbage-collected with all remaining tensors when the container is destroyed during rebalancing.

**Review by justheuristic — APPROVED** [review]
Great work! If possible, please minimize the renames on module_backends

File: src/petals/server/server.py [hunk 3]
```diff
@@ -119,7 +120,6 @@ def __init__(
         visible_maddrs_str = [str(a) for a in self.dht.get_visible_maddrs()]
         if initial_peers == PUBLIC_INITIAL_PEERS:
             logger.info(f"Connecting to the public swarm, peer_id = {self.dht.peer_id}")
-            logger.info("Please check that your server is reachable at http://health.petals.ml")
```
Comment (borzunov) [inline src/petals/server/server.py:122]: Now we show this message only after the server has actually announced JOINING

**Pull request merged**